Support `submitit` Hydra launcher, to launch runs on SLURM clusters

nathanpainchaud opened the pull request:
The use of this launcher, once tested over a few weeks to launch real jobs, is meant to completely replace the [`vital.utils.jobs`](https://github.com/nathanpainchaud/vital/tree/dev/vital/utils/jobs) package, which will eventually be deleted.

nathanpainchaud committed:
Support `submitit` Hydra launcher, to launch runs on SLURM clusters

The use of this launcher is meant in time to completely replace the `vital.utils.jobs` package

ThierryJudge reviewed (commented):
I think this might require some extra documentation (maybe another readme) but looks good. I added a few questions and might add some more when I test it out.

lemairecarl reviewed (commented):
Quelques questions, mais ça me semble bon!

nathanpainchaud commented:
@lemairecarl @ThierryJudge I finally received an answer recently to the issue I had opened with Hydra regarding the composition order of the config, and how built-in nodes are a special case.

The short answer is that Hydra's current behavior is a feature, not a bug, so the previous way I designed to handle switching between local/Slurm runs wouldn't work. Therefore, I changed the design a bit to introduce a new custom `launcher` config group (not to be confused with the built-in `hydra/launcher` config group).

The purpose of the new config group is to "override" in a way the built-in launcher, so that we use it rather than `hydra/launcher` to switch between local/Slurm runs. This is less elegant than before, but it was necessary because, from my discussions with the Hydra dev, we essentially can't override application options from the override of a built-in config group. So I had no choice but to a create new config group (that I could set to be resolved after application-level options), to enable the necessary overrides.

I know it's been a long time (and I'm in no hurry to get this merged), but if you ever have the time I would appreciate if you could give your opinions on the new design. It would be nice to finally have support for launching `vital` jobs on Slurm clusters again :slightly_smiling_face:

nathanpainchaud committed:
Move unsupported parameters from `hydra/launcher` config group to custom `launcher`

This allows us to revert the custom checks on the config that made sure required launcher config fields exist (because now we can rely on Hydra's built-in parsing)

nathanpainchaud committed:
Update default Beluga config to favor generic options where possible
